lib/rucio/db/sqla/models.py

```diff
@@ -579,8 +579,11 @@ class DataIdentifierAssociationHistory(BASE, ModelBase):
     rule_evaluation = Column(Boolean(name='CONTENTS_HIST_RULE_EVAL_CHK'))
     did_created_at = Column(DateTime)
     deleted_at = Column(DateTime)
-    _table_args = (PrimaryKeyConstraint('scope', 'name', 'child_scope', 'child_name', name='CONTENTS_HIST_PK'),
-                   CheckConstraint('DID_TYPE IS NOT NULL', name='CONTENTS_HIST_DID_TYPE_NN'),
+    __mapper_args__ = {
+        'primary_key': [scope, name, child_scope, child_name]
```

Comment:
I suggest to add `deleted_at` or another date/id column in the PK, then the insertion of the same tuple (scope, name, child_scope, child_name) won't fail and will avoid the extra lookup with the advantage to keep the history of operations and favouring performance.

Reply:
ah. It's to micmic the pk creation. not obvious as `primary_key` is misleading but less hacky too :)  I think the lookup in the above query against the dataIdentifierAssociationHistory table can be removed -- didn't check the logic carefully but several entries for the same tuple is fine with  simple bulk inserts vs. selects + bulk inserts (imo).  The non unique index on [scope, name, child_scope, child_name] can be added there too, no ?

Reply:
yes exactly, it's just to mimic to sqlalchemy that there is a PK, while in reality there isn't 😄 I checked the table create queries, like this no PK gets created, but sqlalchemy ORM still works as it thinks these columns are a PK. Unless you don't handle these objects too much (select + update) this should be completely fine.
The (non-unique) index creation we have to discuss. One argument could be that if we only rarely query these tables anyway (or even do it in hadoop) there is no point in "wasting" the space for an index (Which can be considerable for a big history table) and communities who query them often (or don't use hadoop) can create the index themselves. On the other hand we could just create them in the first place, and communities who don't need them, can drop them. Let's discuss this in #1818 😄